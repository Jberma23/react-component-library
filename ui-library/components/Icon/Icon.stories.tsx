import { Meta, StoryFn, StoryObj } from "@storybook/react";
import { ArrowLeftIcon } from "@ui-library/icons";
import * as icons from "@ui-library/index";
import { Icon } from "./Icon";

type IconName = keyof typeof icons;

const meta: Meta<typeof Icon> = {
  component: Icon,
  title: "Library/Icon",
  args: {
    size: "default",
    SvgIcon: ArrowLeftIcon,
  },
  argTypes: {
    SvgIcon: {
      table: {
        disable: true,
      },
      control: "select",
      options: ["foo", "bar"],
    },
  },
  parameters: { backgrounds: { default: "light" } },
};
export default meta;

type Story = StoryObj<typeof Icon>;

export const SizeDefault: Story = {
  args: {
    size: "default",
  },
};

export const SizeSmall: Story = {
  args: {
    size: "small",
  },
};
export const SizeLarge: Story = {
  args: { size: "large" },
};

export const FillByCurrentColor: StoryFn<typeof Icon> = (args) => (
  <div className="text-primary">
    <Icon {...args} className="mr-xs" size={"small"} />
    Primary color applied in container
  </div>
);

export const IconGrid: StoryFn = () => (
  <div className="grid grid-cols-3 lg:grid-cols-5">
    {Object.keys(icons).map((iconName, i) => (
      <div key={i} className=" flex flex-col gap-s items-center m-m">
        <Icon
          SvgIcon={icons[iconName as IconName]}
          size="large"
          title={iconName}
        />
        <p>{iconName}</p>
      </div>
    ))}
  </div>
);
