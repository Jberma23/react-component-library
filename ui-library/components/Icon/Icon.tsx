import { cva } from "cva";

const iconSizes = {
  default: 24,
  small: 16,
  large: 32,
  scale: undefined,
};

export type IconSize = keyof typeof iconSizes;

export interface IconProps extends React.SVGProps<SVGSVGElement> {
  SvgIcon: React.ComponentType<React.SVGProps<SVGSVGElement>>;
  size?: IconSize;
}

const iconStyles = cva(["fill-current inline"]);

export const Icon = ({ SvgIcon, size, className, ...rest }: IconProps) => {
  const iconSize = iconSizes[size || "default"];
  return (
    <SvgIcon
      className={iconStyles({ className })}
      width={iconSize}
      height={iconSize}
      {...rest}
    />
  );
};
