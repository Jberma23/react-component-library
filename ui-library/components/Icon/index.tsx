// istanbul ignore file
export { Icon, type IconProps } from "./Icon";
