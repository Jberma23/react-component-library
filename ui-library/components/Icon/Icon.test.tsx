import { render, screen } from "@testing-library/react";
import { UserIcon } from "@ui-library/index";
import { Icon } from "./Icon";

describe(Icon, () => {
  it("renders with defaults", () => {
    const { asFragment } = render(<Icon SvgIcon={UserIcon} title="User" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with non-default size", () => {
    const { asFragment } = render(
      <Icon SvgIcon={UserIcon} size="small" title="User" />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with passed class name", () => {
    const testClassName = "test";
    render(
      <Icon
        SvgIcon={UserIcon}
        className={testClassName}
        data-testid="icon-test"
        title="User"
      />
    );
    expect(screen.getByTestId("icon-test")).toHaveClass(testClassName);
  });
});
