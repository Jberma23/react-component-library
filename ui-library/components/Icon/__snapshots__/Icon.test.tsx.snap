// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders with defaults 1`] = `
<DocumentFragment>
  <svg
    aria-hidden="true"
    class="fill-current inline"
    fill="currentColor"
    height="24"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      User
    </title>
    <path
      clip-rule="evenodd"
      d="M7.5 6a4.5 4.5 0 119 0 4.5 4.5 0 01-9 0zM3.751 20.105a8.25 8.25 0 0116.498 0 .75.75 0 01-.437.695A18.683 18.683 0 0112 22.5c-2.786 0-5.433-.608-7.812-1.7a.75.75 0 01-.437-.695z"
      fill-rule="evenodd"
    />
  </svg>
</DocumentFragment>
`;

exports[`Icon renders with non-default size 1`] = `
<DocumentFragment>
  <svg
    aria-hidden="true"
    class="fill-current inline"
    fill="currentColor"
    height="16"
    viewBox="0 0 24 24"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      User
    </title>
    <path
      clip-rule="evenodd"
      d="M7.5 6a4.5 4.5 0 119 0 4.5 4.5 0 01-9 0zM3.751 20.105a8.25 8.25 0 0116.498 0 .75.75 0 01-.437.695A18.683 18.683 0 0112 22.5c-2.786 0-5.433-.608-7.812-1.7a.75.75 0 01-.437-.695z"
      fill-rule="evenodd"
    />
  </svg>
</DocumentFragment>
`;
